File: diagram-poc/frontend/src/app/core/services/designwin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { apiBaseUrl } from '../app-config';

/**
 * Arrow Design Win data, proxied by the backend (/api/designwin/*) so the OAuth
 * credentials stay server-side. Responses are the raw Design Win JSON; the
 * panel normalises them tolerantly since field names vary across environments.
 */
@Injectable({ providedIn: 'root' })
export class DesignWinService {
  private readonly api = apiBaseUrl();

  constructor(private http: HttpClient) {}

  private get(path: string, params: Record<string, string | undefined>): Observable<any> {
    let hp = new HttpParams();
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== null && String(v).trim() !== '') hp = hp.set(k, String(v).trim());
    }
    return this.http.get<any>(`${this.api}/designwin/${path}`, { params: hp });
  }

  customers(customerName?: string, billToNumber?: string): Observable<any> {
    return this.get('customers', { customerName, billToNumber });
  }

  projects(customerName?: string, projectName?: string, billToNumber?: string): Observable<any> {
    return this.get('projects', { customerName, projectName, billToNumber });
  }

  boards(projectId?: string, projectName?: string): Observable<any> {
    return this.get('boards', { projectId, projectName });
  }

  registrationDetails(opts: {
    arrowUniqueNum?: string; registrationNum?: string; boardNum?: string; trackingNum?: string;
  }): Observable<any> {
    return this.get('registration-details', opts);
  }

  custParts(opts: {
    customerName?: string; custBillTo?: string; projectId?: string; boardNum?: string; projectName?: string;
  }): Observable<any> {
    return this.get('cust-parts', opts);
  }

  /** POS / sales history for a part — "field-proven" signal. */
  sales(partNumber: string, mfrName?: string): Observable<any> {
    return this.get('sales', { partNumber, mfrName });
  }
}

File: diagram-poc/backend/src/main/java/com/example/diagram/service/DesignWinService.java
package com.example.diagram.service;

/**
 * Proxies the Arrow Design Win APIs (customers, projects, boards, registration
 * details, customer part search and POS/sales). Each method returns the raw JSON
 * body from Arrow so the frontend can render it directly. Required-parameter
 * validation is enforced here; transport/auth is handled by the shared client.
 */
public interface DesignWinService {

    /** Customer / BillTo search. Requires customerName or billToNumber. */
    String customers(String customerName, String billToNumber, String operatingUnit);

    /** Project search. Requires one of customerName / projectName / billToNumber. */
    String projects(String customerName, String projectName, String billToNumber);

    /** Board search. Requires projectId or projectName. */
    String boards(String projectId, String projectName);

    /** Registration details (AUN search). Requires one identifier. */
    String registrationDetails(String arrowUniqueNum, String registrationNum,
                               String boardNum, String trackingNum);

    /** Customer part search. Requires customerName or custBillTo. */
    String custPartSearch(String customerName, String custBillTo,
                          String projectId, String boardNum, String projectName);

    /** POS / sales lookup. Requires partNumber. */
    String sales(String partNumber, String mfrName);
}
